File: repos/openjarvis/src/openjarvis/agents/native_openhands.py
"""NativeOpenHandsAgent -- code-execution-centric agent.

Renamed from ``OpenHandsAgent`` to clarify this is OpenJarvis's native
CodeAct-style implementation.  The ``OpenHandsAgent`` name is now used
for the real openhands-sdk integration in ``openhands.py``.
"""

from __future__ import annotations

import json as _json
import re
from typing import Any, List, Optional

from openjarvis.agents._stubs import AgentContext, AgentResult, ToolUsingAgent
from openjarvis.core.events import EventBus
from openjarvis.core.registry import AgentRegistry
from openjarvis.core.types import Message, Role, ToolCall, ToolResult
from openjarvis.engine._stubs import InferenceEngine
from openjarvis.tools._stubs import BaseTool, build_tool_descriptions

OPENHANDS_SYSTEM_PROMPT = (  # noqa: E501
    "You are an AI assistant with access to tools. "
    "You MUST use tools when they would help answer "
    "the user's question.\n\n"
    "## How to use tools\n\n"
    "To call a tool, write on its own lines:\n\n"
    "Action: <tool_name>\n"
    "Action Input: <json_arguments>\n\n"
    "You will receive the result, then continue your "
    "response.\n\n"
    "## Available tools\n\n"
    "{tool_descriptions}\n\n"
    "## Important rules\n\n"
    "- When the user asks you to look up, search, fetch, "
    "or summarize a URL or topic, you MUST use web_search. "
    "Do NOT say you cannot browse the web.\n"
    "- When the user provides a URL, pass the FULL URL "
    "(including https://) as the query to web_search. "
    "Do NOT rewrite URLs into search keywords.\n"
    "- When the user asks a math question, use calculator.\n"
    "- When the user asks to read a file, use file_read.\n"
    "- You CAN write Python code in ```python blocks and "
    "it will be executed. Use this for computation, data "
    "processing, or when no specific tool fits.\n"
    "- If no tool or code is needed, respond directly "
    "with your answer.\n"
    "- Do NOT include <think> tags or internal reasoning "
    "in your response. Respond directly."
)


@AgentRegistry.register("native_openhands")
class NativeOpenHandsAgent(ToolUsingAgent):
    """Native CodeAct agent -- generates and executes Python code."""

    agent_id = "native_openhands"

    def __init__(
        self,
        engine: InferenceEngine,
        model: str,
        *,
        tools: Optional[List[BaseTool]] = None,
        bus: Optional[EventBus] = None,
        max_turns: int = 3,
        temperature: float = 0.7,
        max_tokens: int = 2048,
    ) -> None:
        super().__init__(
            engine, model, tools=tools, bus=bus,
            max_turns=max_turns, temperature=temperature,
            max_tokens=max_tokens,
        )

    @staticmethod
    def _expand_urls(text: str) -> tuple[str, bool]:
        """If the user message contains a URL, fetch it and inline the content.

        Returns (possibly_expanded_text, was_expanded).
        """
        import re as _re

        url_match = _re.search(r"https?://[^\s,;\"'<>]+", text)
        if not url_match:
            return text, False
        url = url_match.group(0).rstrip(".,;)")
        try:
            from openjarvis.tools.web_search import WebSearchTool

            content = WebSearchTool._fetch_url(url, max_chars=4000)
            header = f"\n\n--- Content from {url} ---\n"
            footer = "\n--- End of content ---\n"
            expanded = text.replace(
                url, f"{header}{content}{footer}"
            )
            return expanded, True
        except Exception:
            return text, False

    def _truncate_if_needed(
        self,
        messages: list[Message],
        max_prompt_tokens: int = 3000,
    ) -> list[Message]:
        """Truncate messages if estimated token count exceeds limit."""
        total_chars = sum(len(m.content) for m in messages)
        estimated_tokens = total_chars // 4
        if estimated_tokens <= max_prompt_tokens:
            return messages
        # Find the last user message and truncate its content
        for i in range(len(messages) - 1, -1, -1):
            if messages[i].role == Role.USER:
                excess_tokens = estimated_tokens - max_prompt_tokens
                excess_chars = excess_tokens * 4
                original = messages[i].content
                if len(original) > excess_chars + 200:
                    truncated = original[: len(original) - excess_chars]
                    messages[i] = Message(
                        role=Role.USER,
                        content=(
                            truncated
                            + "\n\n[Input truncated"
                            " to fit context window]"
                        ),
                    )
                break
        return messages

    @staticmethod
    def _strip_tool_call_text(text: str) -> str:
        """Remove raw tool call artifacts from final output."""
        # Remove Action: ... Action Input: ... blocks
        text = re.sub(
            r"Action:\s*.+?(?:Action Input:\s*.+?)?(?=\n\n|\Z)",
            "", text, flags=re.DOTALL | re.IGNORECASE,
        )
        # Remove <tool_call>...</tool_call> or </tool_name> blocks
        text = re.sub(r"<tool_call>.*?</\w+>", "", text, flags=re.DOTALL)
        return text.strip()

    def _extract_code(self, text: str) -> str | None:
        """Extract Python code from markdown code blocks."""
        match = re.search(r"```python\n(.*?)```", text, re.DOTALL)
        if match:
            return match.group(1).strip()
        return None

    def _extract_tool_call(self, text: str) -> tuple[str, str] | None:
        """Extract tool call from structured output.

        Supports two formats:
        1. Action: tool_name / Action Input: {"key": "value"}
        2. <tool_call>tool_name\\n$key=value</tool_call> (XML-style)
        """
        # Format 1: Action / Action Input
        action_match = re.search(r"Action:\s*(.+)", text, re.IGNORECASE)
        input_match = re.search(
            r"Action Input:\s*(.+?)(?=\n\n|\Z)", text, re.DOTALL | re.IGNORECASE
        )
        if action_match:
            return (
                action_match.group(1).strip(),
                input_match.group(1).strip() if input_match else "{}",
            )

        # Format 2: <tool_call>tool_name ... </tool_call> or </tool_name>
        xml_match = re.search(
            r"<tool_call>\s*(\w+)\s*(.*?)</\w+>",
            text,
            re.DOTALL,
        )
        if xml_match:
            tool_name = xml_match.group(1).strip()
            raw_params = xml_match.group(2).strip()
            # Parse $key=value or <key>value</key> params into JSON
            params: dict[str, Any] = {}
            # $key=value format
            pat = r"\$(\w+)=(.+?)(?=\$|\n<|</|$)"
            for m in re.finditer(pat, raw_params, re.DOTALL):
                params[m.group(1)] = m.group(2).strip().rstrip("</>\n")
            # <key>value</key> format
            for m in re.finditer(r"<(\w+)>(.*?)</\1>", raw_params, re.DOTALL):
                key, val = m.group(1), m.group(2).strip()
                # Try to parse as int
                try:
                    params[key] = int(val)
                except ValueError:
                    params[key] = val
            # key: value format (common in GLM models)
            if not params:
                for m in re.finditer(
                    r"(\w+)\s*:\s*(.+?)(?=\n\w+\s*:|$)", raw_params, re.DOTALL
                ):
                    key, val = m.group(1), m.group(2).strip().strip("\"'")
                    try:
                        params[key] = int(val)
                    except ValueError:
                        params[key] = val
            if params:
                return (tool_name, _json.dumps(params))
            return (tool_name, "{}")

        return None

    def run(
        self,
        input: str,
        context: Optional[AgentContext] = None,
        **kwargs: Any,
    ) -> AgentResult:
        self._emit_turn_start(input)

        tool_descriptions = build_tool_descriptions(self._tools)
        system_prompt = OPENHANDS_SYSTEM_PROMPT.format(
            tool_descriptions=tool_descriptions,
        )

        # Pre-fetch any URLs in the input so the LLM gets the content directly
        input, url_expanded = self._expand_urls(input)

        # If URL content was inlined, skip the tool loop -- just summarize directly
        if url_expanded:
            direct_messages: list[Message] = [
                Message(
                    role=Role.SYSTEM,
                    content=(
                        "You are a helpful assistant. "
                        "Respond directly to the user's "
                        "request using the provided content."
                        " Do NOT include <think> tags."
                    ),
                ),
                Message(role=Role.USER, content=input),
            ]
            direct_messages = self._truncate_if_needed(direct_messages)
            try:
                result = self._generate(direct_messages)
                content = self._strip_think_tags(result.get("content", ""))
                usage = result.get("usage", {})
                self._emit_turn_end(turns=1)
                return AgentResult(
                    content=content, tool_results=[], turns=1,
                    metadata={
                        "prompt_tokens": usage.get("prompt_tokens", 0),
                        "completion_tokens": usage.get("completion_tokens", 0),
                        "total_tokens": usage.get("total_tokens", 0),
                    },
                )
            except Exception as exc:
                error_str = str(exc)
                if "400" in error_str:
                    error_msg = (
                        "The input is too long for the "
                        "model's context window. "
                        "Please try a shorter message."
                    )
                else:
                    error_msg = (
                        "The model returned an error: "
                        + error_str
                    )
                self._emit_turn_end(turns=1, error=True)
                return AgentResult(
                    content=error_msg,
                    tool_results=[],
                    turns=1,
                    metadata={"error": True},
                )

        messages = self._build_messages(input, context, system_prompt=system_prompt)
        messages = self._truncate_if_needed(messages)

        all_tool_results: list[ToolResult] = []
        turns = 0
        last_content = ""
        total_usage = {"prompt_tokens": 0, "completion_tokens": 0, "total_tokens": 0}

        for _turn in range(self._max_turns):
            turns += 1
            # Truncate before every generate call -- tool results may have
            # expanded the context beyond what the model supports.
            messages = self._truncate_if_needed(messages)

            try:
                result = self._generate(messages)
            except Exception as exc:
                error_str = str(exc)
                if "400" in error_str:
                    error_msg = (
                        "The input is too long for the model's context window. "
                        "Please try a shorter message."
                    )
                else:
                    error_msg = f"The model returned an error: {error_str}"
                self._emit_turn_end(turns=turns, error=True)
                return AgentResult(
                    content=error_msg,
                    tool_results=all_tool_results,
                    turns=turns,
                    metadata={"error": True},
                )

            # Accumulate usage from this generate call
            usage = result.get("usage", {})
            for k in total_usage:
                total_usage[k] += usage.get(k, 0)

            content = result.get("content", "")
            # Strip think tags so they don't interfere with parsing
            content = self._strip_think_tags(content)
            last_content = content

            # Try to extract code
            code = self._extract_code(content)
            if code:
                messages.append(Message(role=Role.ASSISTANT, content=content))

                # Execute via code_interpreter tool if available
                tool_call = ToolCall(
                    id=f"code_{turns}",
                    name="code_interpreter",
                    arguments=_json.dumps({"code": code}),
                )
                tool_result = self._executor.execute(tool_call)
                all_tool_results.append(tool_result)

                obs_text = tool_result.content
                if len(obs_text) > 4000:
                    obs_text = obs_text[:4000] + "\n\n[Output truncated]"
                observation = f"Output:\n{obs_text}"
                messages.append(Message(role=Role.USER, content=observation))
                continue

            # Try tool call
            tool_info = self._extract_tool_call(content)
            if tool_info:
                action, action_input = tool_info
                messages.append(Message(role=Role.ASSISTANT, content=content))

                tool_call = ToolCall(
                    id=f"tool_{turns}", name=action, arguments=action_input
                )
                tool_result = self._executor.execute(tool_call)
                all_tool_results.append(tool_result)

                obs_text = tool_result.content
                if len(obs_text) > 4000:
                    obs_text = obs_text[:4000] + "\n\n[Output truncated]"
                observation = f"Result: {obs_text}"
                messages.append(Message(role=Role.USER, content=observation))
                continue

            # No code or tool call -- this is the final answer
            content = self._strip_think_tags(content)
            content = self._strip_tool_call_text(content)
            self._emit_turn_end(turns=turns)
            return AgentResult(
                content=content, tool_results=all_tool_results, turns=turns,
                metadata=total_usage,
            )

        # Max turns
        final = self._strip_think_tags(last_content) or "Maximum turns reached."
        final = self._strip_tool_call_text(final)
        result = self._max_turns_result(all_tool_results, turns, content=final)
        result.metadata.update(total_usage)
        return result


__all__ = ["NativeOpenHandsAgent"]

File: rust/crates/openjarvis-agents/src/native_openhands.rs
//! NativeOpenHandsAgent -- CodeAct-style agent that uses code actions.
//!
//! Generates and dispatches code (Python blocks) and tool calls to accomplish
//! tasks.  Mirrors the Python ``NativeOpenHandsAgent`` which supports both
//! ``Action: / Action Input:`` structured tool calls and fenced
//! ````python`` code blocks executed via the ``code_interpreter`` tool.

use crate::loop_guard::LoopGuard;
use crate::traits::OjAgent;
use crate::utils::strip_think_tags;
use openjarvis_core::{AgentContext, AgentResult, OpenJarvisError, ToolResult};
use openjarvis_tools::executor::ToolExecutor;
use regex::Regex;
use rig::agent::AgentBuilder;
use rig::completion::message::Message as RigMessage;
use rig::completion::request::{Chat, CompletionModel};
use std::collections::HashMap;
use std::sync::Arc;

// ---------------------------------------------------------------------------
// System prompt
// ---------------------------------------------------------------------------

const OPENHANDS_SYSTEM_PROMPT: &str = "\
You are an AI assistant with access to tools. \
You MUST use tools when they would help answer the user's question.

## How to use tools

To call a tool, write on its own lines:

Action: <tool_name>
Action Input: <json_arguments>

You will receive the result, then continue your response.

## Available tools

{tool_list}

## Important rules

- When the user asks you to look up, search, fetch, or summarize a URL or \
topic, you MUST use web_search. Do NOT say you cannot browse the web.
- When the user provides a URL, pass the FULL URL (including https://) as the \
query to web_search. Do NOT rewrite URLs into search keywords.
- When the user asks a math question, use calculator.
- When the user asks to read a file, use file_read.
- You CAN write Python code in ```python blocks and it will be executed. Use \
this for computation, data processing, or when no specific tool fits.
- If no tool or code is needed, respond directly with your answer.
- Do NOT include <think> tags or internal reasoning in your response. Respond \
directly.";

// ---------------------------------------------------------------------------
// Agent implementation
// ---------------------------------------------------------------------------

/// Native CodeAct agent -- generates and executes code actions (shell commands,
/// file edits) and structured tool calls to accomplish tasks.
///
/// Supports two action formats:
/// 1. `Action: tool_name` / `Action Input: {json}` -- dispatched to the
///    `ToolExecutor`.
/// 2. Fenced ````python` code blocks -- dispatched to the `code_interpreter`
///    tool.
pub struct NativeOpenHandsAgent<M: CompletionModel> {
    agent: rig::agent::Agent<M>,
    executor: Arc<ToolExecutor>,
    max_turns: usize,
}

impl<M: CompletionModel> NativeOpenHandsAgent<M> {
    pub fn new(
        model: M,
        executor: Arc<ToolExecutor>,
        max_turns: usize,
        temperature: f64,
    ) -> Self {
        let tool_list = executor.list_tools().join(", ");
        let system_prompt = OPENHANDS_SYSTEM_PROMPT.replace("{tool_list}", &tool_list);

        let agent = AgentBuilder::new(model)
            .preamble(&system_prompt)
            .temperature(temperature)
            .build();

        Self {
            agent,
            executor,
            max_turns,
        }
    }

    /// Parse `Action:` and `Action Input:` lines from model output.
    fn parse_action(text: &str) -> Option<(String, String)> {
        let action_re = Regex::new(r"(?mi)^Action:\s*(.+)$").unwrap();
        let input_re = Regex::new(r"(?mi)^Action Input:\s*(.+?)(?:\n\n|\z)").unwrap();

        let action = action_re
            .captures(text)?
            .get(1)?
            .as_str()
            .trim()
            .to_string();
        let input = input_re
            .captures(text)
            .and_then(|c| c.get(1))
            .map(|m| m.as_str().trim().to_string())
            .unwrap_or_else(|| "{}".to_string());

        Some((action, input))
    }

    /// Extract Python code from fenced ````python` blocks.
    fn extract_code(text: &str) -> Option<String> {
        let re = Regex::new(r"(?s)```python\n(.*?)```").unwrap();
        re.captures(text)
            .and_then(|c| c.get(1))
            .map(|m| m.as_str().trim().to_string())
    }

    /// Remove raw tool-call artifacts from final output text.
    fn strip_tool_call_text(text: &str) -> String {
        // Remove Action: ... Action Input: ... blocks
        let action_re =
            Regex::new(r"(?si)Action:\s*.+?(?:Action Input:\s*.+?)?(?:\n\n|\z)").unwrap();
        let cleaned = action_re.replace_all(text, "");
        // Remove <tool_call>...</tool_name> XML blocks
        let xml_re = Regex::new(r"(?s)<tool_call>.*?</\w+>").unwrap();
        let cleaned = xml_re.replace_all(&cleaned, "");
        cleaned.trim().to_string()
    }

    /// Truncate observation text if it exceeds 4000 characters.
    fn truncate_observation(content: &str, limit: usize) -> String {
        if content.len() > limit {
            let mut truncated = content[..limit].to_string();
            truncated.push_str("\n\n[Output truncated]");
            truncated
        } else {
            content.to_string()
        }
    }
}

#[async_trait::async_trait]
impl<M: CompletionModel + 'static> OjAgent for NativeOpenHandsAgent<M> {
    fn agent_id(&self) -> &str {
        "native_openhands"
    }

    fn accepts_tools(&self) -> bool {
        true
    }

    async fn run(
        &self,
        input: &str,
        context: Option<&AgentContext>,
    ) -> Result<AgentResult, OpenJarvisError> {
        let mut history: Vec<RigMessage> = context
            .map(|ctx| {
                ctx.conversation
                    .messages
                    .iter()
                    .filter_map(|m| match m.role {
                        openjarvis_core::Role::User => {
                            Some(RigMessage::user(&m.content))
                        }
                        openjarvis_core::Role::Assistant => {
                            Some(RigMessage::assistant(&m.content))
                        }
                        _ => None,
                    })
                    .collect()
            })
            .unwrap_or_default();

        let mut all_tool_results: Vec<ToolResult> = Vec::new();
        let mut guard = LoopGuard::default();
        let mut current_input = input.to_string();

        for turn in 1..=self.max_turns {
            let response = self
                .agent
                .chat(&current_input, history.clone())
                .await
                .map_err(|e| {
                    OpenJarvisError::Agent(openjarvis_core::error::AgentError::Execution(
                        e.to_string(),
                    ))
                })?;

            let text = strip_think_tags(&response);

            // 1. Try to extract a Python code block -> execute via code_interpreter
            if let Some(code) = Self::extract_code(&text) {
                let tool_name = "code_interpreter";
                let args = serde_json::json!({"code": code});
                let args_str = args.to_string();

                if let Some(loop_msg) = guard.check(tool_name, &args_str) {
                    return Ok(AgentResult {
                        content: format!("Agent stopped: {}", loop_msg),
                        tool_results: all_tool_results,
                        turns: turn,
                        metadata: HashMap::new(),
                    });
                }

                let tool_result = match self.executor.execute(
                    tool_name,
                    &args,
                    Some("native_openhands"),
                    None,
                ) {
                    Ok(r) => r,
                    Err(e) => ToolResult::failure(tool_name, e.to_string()),
                };

                let obs = Self::truncate_observation(&tool_result.content, 4000);
                history.push(RigMessage::assistant(&text));
                current_input = format!("Output:\n{}", obs);

                all_tool_results.push(tool_result);
                continue;
            }

            // 2. Try to extract a structured tool call (Action: / Action Input:)
            if let Some((action, action_input)) = Self::parse_action(&text) {
                if let Some(loop_msg) = guard.check(&action, &action_input) {
                    return Ok(AgentResult {
                        content: format!("Agent stopped: {}", loop_msg),
                        tool_results: all_tool_results,
                        turns: turn,
                        metadata: HashMap::new(),
                    });
                }

                let params: serde_json::Value =
                    serde_json::from_str(&action_input).unwrap_or(serde_json::json!({}));

                let tool_result = match self.executor.execute(
                    &action,
                    &params,
                    Some("native_openhands"),
                    None,
                ) {
                    Ok(r) => r,
                    Err(e) => ToolResult::failure(&action, e.to_string()),
                };

                let obs = Self::truncate_observation(&tool_result.content, 4000);
                history.push(RigMessage::assistant(&text));
                current_input = format!("Result: {}", obs);

                all_tool_results.push(tool_result);
                continue;
            }

            // 3. No code or tool call -- this is the final answer
            let cleaned = Self::strip_tool_call_text(&text);
            return Ok(AgentResult {
                content: cleaned,
                tool_results: all_tool_results,
                turns: turn,
                metadata: HashMap::new(),
            });
        }

        // Max turns exceeded
        Ok(AgentResult {
            content: format!("Reached maximum turns ({})", self.max_turns),
            tool_results: all_tool_results,
            turns: self.max_turns,
            metadata: HashMap::new(),
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    use openjarvis_engine::rig_adapter::RigModelAdapter;
    type OpenHandsAgent = NativeOpenHandsAgent<RigModelAdapter<openjarvis_engine::Engine>>;

    #[test]
    fn test_parse_action() {
        let text = "I need to search.\nAction: web_search\nAction Input: {\"query\": \"rust lang\"}";
        let (action, input) = OpenHandsAgent::parse_action(text).unwrap();
        assert_eq!(action, "web_search");
        assert!(input.contains("rust lang"));
    }

    #[test]
    fn test_parse_action_missing_input() {
        let text = "Action: calculator\n\nSome other text";
        let (action, input) = OpenHandsAgent::parse_action(text).unwrap();
        assert_eq!(action, "calculator");
        assert_eq!(input, "{}");
    }

    #[test]
    fn test_extract_code() {
        let text = "Let me compute that:\n```python\nprint(2 + 2)\n```\nDone.";
        let code = OpenHandsAgent::extract_code(text).unwrap();
        assert_eq!(code, "print(2 + 2)");
    }

    #[test]
    fn test_extract_code_none() {
        let text = "No code here, just text.";
        assert!(OpenHandsAgent::extract_code(text).is_none());
    }

    #[test]
    fn test_strip_tool_call_text() {
        let text = "Here is the answer.\nAction: calc\nAction Input: {\"x\": 1}\n\nFinal part.";
        let cleaned = OpenHandsAgent::strip_tool_call_text(text);
        assert!(!cleaned.contains("Action:"));
        assert!(cleaned.contains("Final part"));
    }

    #[test]
    fn test_truncate_observation() {
        let short = "hello";
        assert_eq!(
            OpenHandsAgent::truncate_observation(short, 100),
            "hello"
        );

        let long = "x".repeat(5000);
        let truncated = OpenHandsAgent::truncate_observation(&long, 100);
        assert!(truncated.len() < 200);
        assert!(truncated.contains("[Output truncated]"));
    }
}
